File: src/main/java/mountainrangepvp/game/world/HillsHeightMap.java
package mountainrangepvp.game.world;

/**
 * @author lachlan
 */
public class HillsHeightMap extends AbstractHeightMap {

    private static final int SCALE = 5;
    private static final int WALL_WIDTH = 50;
    private static final int WALL_DISTANCE = 3000;
    private static final int WALL_HEIGHT = 110;
    private final int originalSeed;
    private final int seed;
    private final Noise noise;
    private boolean makeWalls;
    private boolean origin;

    public HillsHeightMap(int originalSeed) {
        this.originalSeed = originalSeed;
        this.seed =
                originalSeed ^
                        ((originalSeed << 2) + 10) ^
                        ((originalSeed << 4) + 883) ^
                        ((originalSeed << 6) + 78) ^
                        ((originalSeed << 8) + 43);
        this.noise = new Noise();

        makeWalls = true;
        origin = false;
    }

    @Override
    public int getSeed() {
        return originalSeed;
    }

    @Override
    public int getSample(int x) {
        if (origin && x == 0) {
            return -10000;
        }

        if (makeWalls && Math.abs(x % WALL_DISTANCE) < WALL_WIDTH) {
            // Wall Generator
            int sample = (int) sample(
                    x / WALL_WIDTH * WALL_WIDTH);
            return sample + WALL_HEIGHT;
        }

        // Regular terrain
        return (int) (0.2f * sample(x)
                + 0.8f * (sample(x - 1) + sample(x - 2) + sample(x - 3)
                + sample(x + 1) + sample(x + 2) + sample(x + 3)) / 6);
    }

    private float sample(float x) {
        float noise = 50;

        x /= SCALE;

        noise += InterpolatedNoise1(x / 800f + 100) * 100000;
        noise += InterpolatedNoise1(x / 70f + 230) * 6000;
        noise += InterpolatedNoise1(x / 300f + 234) * 5000;
        noise += InterpolatedNoise1(x / 70f + 12) * 1000;
        noise += InterpolatedNoise1(x / 50f + 5) * 700;
        noise += InterpolatedNoise1(x / 30f + 32) * 300;
        noise += InterpolatedNoise1(x / 5f + 13) * 100;
        /*
         * noise += InterpolatedNoise1(x / 2f) * 100;
         * noise += InterpolatedNoise1(x / 1f) * 50;
         */

        return noise;
    }

    private float InterpolatedNoise1(float x) {
        int integer_X = (int) x;
        float fractional_X = x - integer_X;

        float v1 = SmoothedNoise1(integer_X);
        float v2 = SmoothedNoise1(integer_X + 1);

        return Interpolate(v1, v2, fractional_X) - 0.9f;
    }

    private float Interpolate(float a, float b, float x) {
        return a * (1 - x) + b * x;
    }

    private float SmoothedNoise1(float x) {
        return noise(x) / 2 + noise(x - 1) / 4 + noise(x + 1) / 4;
    }

    private float noise(float x) {
        return noise.noise(x * seed);
    }
}

File: src/main/scala/mountainrangepvp/core/HillsHeightMap.scala
package mountainrangepvp.core

import java.lang.Float._

import com.badlogic.gdx.math.MathUtils

/**
 * @author lachlan
 */
object HillsHeightMap {
  private val SCALE = 5
  private val WALL_WIDTH = 50
  private val WALL_DISTANCE = 3000
  private val WALL_HEIGHT = 110
}

class HillsHeightMap(seed: Int) extends HeightMap {
  private val noise = new Noise
  private val makeWalls = false
  private val origin = false

  /**
   * Period, Amplitude
   */
  private val octaves: List[(Float, Float)] = List(
    (100f, 1000f),
    (20f, 70f)
  )

  def getSeed = seed

  def getSample(x: Int) = {
    if (origin && x == 0) {
      -10000
    } else if (makeWalls && Math.abs(x % HillsHeightMap.WALL_DISTANCE) < HillsHeightMap.WALL_WIDTH) {
      val sample = sampleAt(x / HillsHeightMap.WALL_WIDTH * HillsHeightMap.WALL_WIDTH).toInt
      sample + HillsHeightMap.WALL_HEIGHT
    } else
      (0.1f * sampleAt(x) +
       0.9f * (sampleAt(x - 5) + sampleAt(x + 5) +
               sampleAt(x - 10) + sampleAt(x + 10) +
               sampleAt(x - 20) + sampleAt(x + 20)) / 6).
      toInt
  }

  private def sampleAt(at: Float) = {
    val x = at / HillsHeightMap.SCALE
    octaves.map { case (period, amplitude) => interpolatedNoiseAt(x / period) * amplitude }.sum
  }

  private def interpolatedNoiseAt(x: Float) = {
    val integer_X = x.toInt
    val fractional_X = if (x >= 0) x - integer_X else 1 + x - integer_X
    if (fractional_X < 0)
      println(x + " " + fractional_X)
    val v1 = smoothedNoiseAt(integer_X)
    val v2 = smoothedNoiseAt(integer_X + 1)
    (MathUtils.lerp(v1, v2, fractional_X) - 0.9f) * x.abs.min(1f)
  }

  private def smoothedNoiseAt(x: Float) =
    noiseAt(x) / 2 + noiseAt(x - 3) / 4 + noiseAt(x + 3) / 4


  private def noiseAt(x: Float) = noise.noise(floatToIntBits(x) ^ seed)
}
